modules/loadbalancer/main.tf

```diff
@@ -1,60 +1,33 @@
-}
-
-data "azurerm_public_ip" "exists" {
-  for_each = { for k, v in var.frontend_ips : k => v if try(v.public_ip_name, null) != null }
-
-  name                = each.value.public_ip_name
-  resource_group_name = try(each.value.public_ip_resource_group, var.resource_group_name)
-}
-
-resource "azurerm_lb" "lb" {
-  name                = var.name
-  resource_group_name = var.resource_group_name
-  location            = var.location
-  sku                 = "Standard"
-  tags                = var.tags
-
-  dynamic "frontend_ip_configuration" {
-    for_each = local.frontend_ips
-    iterator = each
-    content {
-      name                          = each.value.name
-      public_ip_address_id          = each.value.public_ip_address_id
-      subnet_id                     = each.value.subnet_id
-      private_ip_address_allocation = each.value.private_ip_address_allocation
-      private_ip_address            = each.value.private_ip_address_allocation == "Static" ? each.value.private_ip_address : null
-      zones                         = try(each.value.zones, null) == null ? [] : each.value.zones
-      # Azure error message says: "Networking supports zones only for frontendIpconfigurations which reference a subnet."
-      # Therefore availability_zone null has a very different meaning depending whether subnet_id is null or not.
-    }
-  }
-}
-
 locals {
-  # The main input will go here through a sequence of operations to obtain the final result.
+  # Decide how the backend machines access internet. If outbound rules are defined use them instead of the default route.
+  # This is an inbound rule setting, applicable to all inbound rules.
+  disable_outbound_snat = length(var.outbound_rules) > 0
+
 
   # Recalculate the main input map, taking into account whether the boolean condition is true/false.
   frontend_ips = {
     for k, v in var.frontend_ips : k => {
+      name  = k
+      rules = try(v.rules, {})
+    }
+  }
+
+  # Frontend ip configurations, no rules, only PIPs.
+  # The `if` statement is to avoid errors in LB configuration when re-using a PIP created by some other rule.
+  frontend_ip_configurations = {
+    for k, v in merge(var.frontend_ips, var.outbound_rules) : k => {
       name                          = k
       public_ip_address_id          = try(v.create_public_ip, false) ? azurerm_public_ip.this[k].id : try(data.azurerm_public_ip.exists[k].id, null)
       create_public_ip              = try(v.create_public_ip, false)
       zones                         = try(v.zones, null)
       subnet_id                     = try(v.subnet_id, null)
       private_ip_address_allocation = try(v.private_ip_address_allocation, null)
       private_ip_address            = try(v.private_ip_address, null)
-      rules                         = try(v.rules, {})
     }
+    if(
+      try(v.create_public_ip, false)
+      || (!try(v.create_public_ip, false) && !can(merge(var.frontend_ips, var.outbound_rules)[v.public_ip_name]))
```

Comment:
I think you can declare local variable and reuse that part **merge(var.frontend_ips, var.outbound_rules)**.
In if statement in OR ( || ) part is always dependent on the second argument:


```
> (!try(true,false) && !(true))
false
> (!try(false,false) && !(true))
false
> (!try(false,false) && !(false))
true
```
So you can simplify to ``!can(merge(var.frontend_ips, var.outbound_rules)[v.public_ip_name]))``

IMO If you add new local value it will looks cleaner

Reply:
your're right, simplified the bool logic + a local with combined rules was added